llpc/translator/lib/SPIRV/SPIRVReader.cpp

```diff
@@ -1730,6 +1727,38 @@ static AtomicOrdering transMemorySemantics(const SPIRVConstant *const spvMemoryS
   return AtomicOrdering::Monotonic;
 }
 
+// =====================================================================================================================
+// Get memory ordering from SPIR-V opcode and its associated memory mask and scope.
+//
+// @param spvValue : A SPIR-V value.
+// @param isSrcOperand : True for OpLoad or OpCopyMemory source, False for OpStore or OpCopyMemory target.
+template <typename T> AtomicOrdering SPIRVToLLVM::getMemoryOrdering(T *const spvValue, bool isSrcOperand) {
+  unsigned memoryAccessMask = spvValue->getMemoryAccessMask(isSrcOperand);
+
+  if (memoryAccessMask & MemoryAccessNonPrivatePointerKHRMask) {
+    return AtomicOrdering::Monotonic;
+  }
+
+  SPIRVWord spvId = SPIRVID_INVALID;
+  // MakePointerVisible is only valid for OpLoad or OpCopyMemory source
+  if (memoryAccessMask & MemoryAccessMakePointerVisibleKHRMask)
+    spvId = spvValue->getMakeVisibleScope(isSrcOperand);
+  // MakePointerAvailable is only valid for OpStore or OpCopyMemory target
+  if (memoryAccessMask & MemoryAccessMakePointerAvailableKHRMask)
+    spvId = spvValue->getMakeAvailableScope(isSrcOperand);
+
+  if (spvId != SPIRVID_INVALID) {
+    SPIRVConstant *const spvScope = static_cast<SPIRVConstant *>(m_bm->getValue(spvId));
+    const unsigned scope = spvScope->getZExtIntValue();
+    const bool isSystemScope = (scope <= ScopeDevice || scope == ScopeQueueFamilyKHR);
+    if (isSystemScope) {
```

Comment:
s/isSystemScope/isDeviceScope/.

System scope is often understood to mean coherence with the CPU as well.